## **Assignment- Week 5**
**Instructions:** 
- Answer each question on a separate cell, either in a markdown or code cell depending on the question.
- FOr e
- Use comments in your code

## **Dataset Description**
You are presented with a customer booking dataset that contains information about airline bookings, including the number of passengers, sales channel, trip type, purchase lead time, length of stay, flight hour, flight day, route, booking origin, and various flags indicating if the customer wanted extra baggage, preferred seat, or in-flight meals. The dataset also includes the total flight duration and a flag indicating if the booking was completed. 

**Using the data, provide the following answers:**
1. What is the shape of our dataframe and explain what the shape means?
2. How many columns have null values and how many duplicate records are in the dataset?
3. Remove duplicates, rows with null values and the `purchase lead` and `booking origin` columns, what is the new shape of the dataframe?
4. What are the different datatypes we have in the dataset?
5. What are the different `trip types` available to customers and which one is mostly used?
6. What is the most used `sales channel` for booking by customers?
7. Which day had the highest `total number of flights`?
8. How many customers wanted a `preferred seat` and in `flight meals`?
9. Which of the `trip type` had the highest average `flight duration`?
10. Which day of the week had the highest average `flight duration`?

## **Columns description**
**Note:**\
1 = True/Yes\
0 = False/No
- `num_passengers` = number of passengers travelling
- `sales_channel` = sales channel booking was made on
- `trip_type = trip` Type (Round Trip, One Way, Circle Trip)
- `purchase_lead` = number of days between travel date and booking date
- `length_of_stay` = number of days spent at destination
- `flight_hour` = hour (time) of flight departure
- `flight_day` = day of week of flight departure
- `booking_origin` = country from where booking was made
- `wants_extra_baggage` = if the customer wanted extra baggage in the booking
- `wants_preferred_seat` = if the customer wanted a preferred seat in the booking
- `wants_in_flight_meals` = if the customer wanted in-flight meals in the booking
- `flight_duration` = total duration of flight (in hours)
- `booking_complete` = flag indicating if the customer completed the booking

In [4]:
import numpy as np
import pandas as pd

### 1. What is the shape of our dataframe and explain what the shape means?

In [6]:
df = pd.read_csv('customer_bookings.csv') # reading the csv file from the directory 
df.shape #Displaying the shape

(50017, 14)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50017 entries, 0 to 50016
Data columns (total 14 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   num_passengers         50017 non-null  int64  
 1   sales_channel          50007 non-null  object 
 2   trip_type              50012 non-null  object 
 3   purchase_lead          49992 non-null  float64
 4   length_of_stay         50010 non-null  float64
 5   flight_hour            50002 non-null  float64
 6   flight_day             50013 non-null  object 
 7   route                  50014 non-null  object 
 8   booking_origin         50006 non-null  object 
 9   wants_extra_baggage    50013 non-null  float64
 10  wants_preferred_seat   50012 non-null  float64
 11  wants_in_flight_meals  50012 non-null  float64
 12  flight_duration        50004 non-null  float64
 13  booking_complete       50017 non-null  int64  
dtypes: float64(7), int64(2), object(5)
memory usage: 5.3+ 

#### The shape of our DataFrame is (50017, 14).

###### This shape means the following:

- 50017 represents the number of rows in the DataFrame. Each row corresponds to a single entry, record, or example within our dataset. A row is a distinct instance we are analyzing.

- 14 represents the number of columns in the DataFrame. Each column corresponds to a different feature, attribute, or variable that describes each of the 50,017 entries. These are the characteristics we have information about for each example.

### 2. How many columns have null values and how many duplicate records are in the dataset?

In [10]:
# Checking for null values in each column and summing them
null_counts_per_column = df.isnull().sum()

# Filtering for columns that have at least one null value
columns_with_nulls = null_counts_per_column[null_counts_per_column > 0]

# Getting the number of columns with null values
num_columns_with_nulls = columns_with_nulls.shape[0]

print(f"The Number of columns with null values are: {num_columns_with_nulls}") # printing the number of columns with null values
print("\nNumber of null values per column:") 
print(columns_with_nulls) # printing the columns with null values and the number of nulls

The Number of columns with null values are: 12

Number of null values per column:
sales_channel            10
trip_type                 5
purchase_lead            25
length_of_stay            7
flight_hour              15
flight_day                4
route                     3
booking_origin           11
wants_extra_baggage       4
wants_preferred_seat      5
wants_in_flight_meals     5
flight_duration          13
dtype: int64


In [11]:
# Identifying duplicate rows
duplicate_rows = df.duplicated()

# Counting the number of duplicate rows
num_duplicate_rows = duplicate_rows.sum()

# printing the number of duplicates on the screen
print(f"The Number of duplicate records in the dataset: {num_duplicate_rows}")

The Number of duplicate records in the dataset: 735


### 3. Remove duplicates, rows with null values and the purchase lead and booking origin columns, what is the new shape of the dataframe?

In [13]:
# Removing duplicate rows
df_no_duplicates = df.drop_duplicates()

# Removing rows with any null values
df_cleaned = df_no_duplicates.dropna()

# Removing the 'purchase_lead' and 'booking_origin' columns
columns_to_drop = ['purchase_lead', 'booking_origin']
df_final = df_cleaned.drop(columns=columns_to_drop, axis = 1)

# Get the new shape of the DataFrame
new_shape = df_final.shape

print(f"Original shape of the DataFrame: {df.shape}")
print(f"Shape after removing duplicates: {df_no_duplicates.shape}")
print(f"Shape after removing rows with null values: {df_cleaned.shape}")
print(f"Shape after removing 'purchase_lead' and 'booking_origin' columns: {new_shape}")

Original shape of the DataFrame: (50017, 14)
Shape after removing duplicates: (49282, 14)
Shape after removing rows with null values: (49207, 14)
Shape after removing 'purchase_lead' and 'booking_origin' columns: (49207, 12)


### 4. What are the different datatypes we have in the dataset?

In [15]:
# Getting the data type of each column
column_dtypes = df.dtypes

# Getting the unique data types present in the Series of column data types
unique_dtypes = column_dtypes.unique()

print("The different data types in the dataset are:") 
for dtype in unique_dtypes:
    print(dtype) # printing the different datatypes

The different data types in the dataset are:
int64
object
float64


### 5. What are the different trip types available to customers and which one is mostly used?

In [17]:
# Getting the unique trip types
unique_trip_types = df['trip_type'].unique()

# Getting the count of each trip type
trip_type_counts = df['trip_type'].value_counts()

# Finding the most frequently used trip type
most_used_trip_type = trip_type_counts.idxmax()
most_used_trip_count = trip_type_counts.max()

print("The different trip types available are:")
print(unique_trip_types)

print("\nUsage count for each trip type are:")
print(trip_type_counts)

print(f"\nThe most frequently used trip type is: '{most_used_trip_type}' with {most_used_trip_count} bookings.")

The different trip types available are:
['RoundTrip' 'CircleTrip' 'OneWay' nan]

Usage count for each trip type are:
trip_type
RoundTrip     49509
OneWay          387
CircleTrip      116
Name: count, dtype: int64

The most frequently used trip type is: 'RoundTrip' with 49509 bookings.


### 6. What is the most used sales channel for booking by customers?

In [19]:
# Getting the count of each sales channel
sales_channel_counts = df['sales_channel'].value_counts()

# Finding the most frequently used sales channel
most_used_sales_channel = sales_channel_counts.idxmax()
most_used_sales_channel_count = sales_channel_counts.max()

print("Usage count for each sales channel:")
print(sales_channel_counts)

print(f"\nThe most frequently used sales channel for booking is: '{most_used_sales_channel}' with {most_used_sales_channel_count} bookings.")

Usage count for each sales channel:
sales_channel
Internet    44386
Mobile       5621
Name: count, dtype: int64

The most frequently used sales channel for booking is: 'Internet' with 44386 bookings.


### 7. Which day had the highest total number of flights?

In [22]:
# Getting the count of flights for each day
flight_day_counts = df['flight_day'].value_counts()

# Finding the day with the highest number of flights
day_with_highest_flights = flight_day_counts.idxmax()
highest_flight_count = flight_day_counts.max()

# Mapping numerical flight_day to day names
day_mapping = {
    1: 'Monday',
    2: 'Tuesday',
    3: 'Wednesday',
    4: 'Thursday',
    5: 'Friday',
    6: 'Saturday',
    7: 'Sunday'
} 

# Getting the day name corresponding to the numerical value
day_name_highest_flights = day_mapping.get(day_with_highest_flights, f'Day {day_with_highest_flights}')

print("Number of flights per day:")
print(flight_day_counts)

print(f"\nThe day with the highest total number of flights was: '{day_name_highest_flights}' with {highest_flight_count} flights.")

Number of flights per day:
flight_day
Mon    8105
Tue    7676
Wed    7674
Thu    7426
Fri    6764
Sun    6556
Sat    5812
Name: count, dtype: int64

The day with the highest total number of flights was: 'Day Mon' with 8105 flights.


### 8. How many customers wanted a preferred seat and in flight meals?

In [25]:
# Finding customers who wanted BOTH preferred seat AND in-flight meals
both_services = df[(df['wants_preferred_seat'] == True) & (df['wants_in_flight_meals'] == True)]

# Getting the number of such customers
num_both_services = len(both_services)

print(f"Number of customers who wanted both a preferred seat and in-flight meals are: {num_both_services}")

Number of customers who wanted both a preferred seat and in-flight meals are: 9908


### 9. Which of the trip type had the highest average flight duration?

In [28]:
# Calculating the average flight duration for each trip type
average_duration_per_trip = df.groupby('trip_type')['flight_duration'].mean()

# Finding the trip type with the highest average flight duration
highest_avg_duration_trip = average_duration_per_trip.idxmax()
highest_average_duration = average_duration_per_trip.max()

print("Average flight duration per trip type:")
print(average_duration_per_trip)

print(f"\nThe trip type with the highest average flight duration is: '{highest_avg_duration_trip}' with an average duration of {highest_average_duration:.2f}.")

Average flight duration per trip type:
trip_type
CircleTrip    7.764828
OneWay        7.966770
RoundTrip     7.271532
Name: flight_duration, dtype: float64

The trip type with the highest average flight duration is: 'OneWay' with an average duration of 7.97.


### 10. Which day of the week had the highest average flight duration?

In [33]:
# Checking for null values in 'flight_duration'
print("Number of null values in 'flight_duration':", df['flight_duration'].isnull().sum())

# Checking the unique values in 'flight_day'
print("Unique values in 'flight_day':", df['flight_day'].unique())

# Calculating the average flight duration for each flight_day
average_duration_per_day = df.groupby('flight_day')['flight_duration'].mean()

print("\nAverage flight duration per flight_day (before dropping NaN):")
print(average_duration_per_day)

# Droping NaN values from the resulting Series if any exist
average_duration_per_day_cleaned = average_duration_per_day.dropna()

print("\nAverage flight duration per flight_day (after dropping NaN):")
print(average_duration_per_day_cleaned)

# Checking if the cleaned Series is empty
if average_duration_per_day_cleaned.empty:
    print("\nError: The 'average_duration_per_day_cleaned' Series is empty. "
          "This suggests there are no valid flight durations for any of the flight days.")
else:
    # Finding the day with the highest average flight duration
    highest_avg_duration_day = average_duration_per_day_cleaned.idxmax()
    highest_average_duration = average_duration_per_day_cleaned.max()

    print(f"\nThe day of the week with the highest average flight duration is: '{highest_avg_duration_day}' with an average duration of {highest_average_duration:.2f}.")

Number of null values in 'flight_duration': 13
Unique values in 'flight_day': ['Sat' 'Wed' 'Thu' 'Mon' 'Sun' 'Tue' 'Fri' nan]

Average flight duration per flight_day (before dropping NaN):
flight_day
Fri    7.249553
Mon    7.264899
Sat    7.302901
Sun    7.374796
Thu    7.250244
Tue    7.258524
Wed    7.260923
Name: flight_duration, dtype: float64

Average flight duration per flight_day (after dropping NaN):
flight_day
Fri    7.249553
Mon    7.264899
Sat    7.302901
Sun    7.374796
Thu    7.250244
Tue    7.258524
Wed    7.260923
Name: flight_duration, dtype: float64

The day of the week with the highest average flight duration is: 'Sun' with an average duration of 7.37.
